packages/database/test/helpers/syncpoint-util.ts

```diff
@@ -0,0 +1,603 @@
+        change.childName
+      );
+      const index = query._queryParams.getIndex();
+      snap = new DataSnapshot(change.snapshotNode, childRef, index);
+    }
+    return new DataEvent(change.type, this, snap, change.prevName);
+  }
+  createCancelEvent(error: Error, path: Path): CancelEvent | null {
+    if (this.eventListener_.hasCancelCallback) {
+      return new CancelEvent(this, error, path);
+    } else {
+      return null;
+    }
+  }
+  getEventRunner(eventData: DataEvent): () => void {
+    const self = this;
+    return function myEventRunner() {
+      return self.eventListener_.onValue(
+        eventData.snapshot,
+        eventData.prevName
+      );
+    };
+  }
+
+  matches(otherEventRegistration: EventRegistration) {
+    if (otherEventRegistration instanceof TestEventRegistration) {
+      // TODO(mtewas unknown as TestEventRegistration rit
+      return (
+        (this as unknown as TestEventRegistration) === otherEventRegistration
+      );
+    }
+    return false;
+  }
+  hasAnyCallback(): boolean {
+    return true;
+  }
+}
+
+function patchFakeAuthFunctions(app: FirebaseApp) {
+  const token_ = null;
+
+  app['INTERNAL'] = app['INTERNAL'] || {};
+
+  app['INTERNAL']['getToken'] = function (forceRefresh) {
+    return Promise.resolve(token_);
+  };
+
+  app['INTERNAL']['addAuthTokenListener'] = function (listener) {};
+
+  app['INTERNAL']['removeAuthTokenListener'] = function (listener) {};
+
+  return app;
+}
+
+class SyncPointListenProvider implements ListenProvider {
+  private listens_: { [key: string]: boolean };
+  constructor() {
+    this.listens_ = {};
+  }
+  startListening(
+    query: QueryContext,
+    tag: number | null,
+    hashFn: () => string,
+    onComplete: (a: string, b?: unknown) => Event[]
+  ): Event[] {
+    const queryParams = query._queryIdentifier;
+    const path = query._path;
+    const logTag = queryParams + (tag ? ' - (' + tag + ')' : '');
+    const key = this.getQueryKey(query, tag);
+    const existing = this.listens_[key];
+    assert(!existing, 'Duplicate listen');
+    this.listens_[key] = true;
+    return [];
+  }
+  stopListening(query: QueryContext, tag: number | null) {
+    const queryParams = query._queryIdentifier;
+    const path = query._path;
+    const logTag = queryParams + (tag ? ' - (' + tag + ')' : '');
+    const key = this.getQueryKey(query, tag);
+    const existing = this.listens_[key];
+    assert(existing, "Missing record of query that we're removing");
+    delete this.listens_[key];
+  }
+  getQueryKey(query: QueryContext, tag: number | null) {
+    return (
+      query._path.toString() +
+      '|' +
+      query._queryIdentifier +
+      (tag ? '|' + tag : '')
+    );
+  }
+}
+
+export class SyncPointTestParser {
+  app: FirebaseApp;
+  listens_: any = {};
+  listenProvider_: ListenProvider;
+  private syncTree_: SyncTree;
+  constructor() {
+    this.app = initializeApp(
+      { databaseURL: 'http://tests.fblocal.com:9000' },
```

Comment:
It looks like these run against an emulator. For the future, it'd be nice if we could point them at prod/staging as well.